src/Symfony/Component/Console/Command/CompleteCommand.php

```diff
@@ -0,0 +1,136 @@
+use Symfony\Component\Console\Exception\ExceptionInterface;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\OutputInterface;
+
+/**
+ * Responsible for providing the values to the shell completion.
+ *
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ */
+class CompleteCommand extends Command
+{
+    protected static $defaultName = '|_complete';
+    protected static $defaultDescription = 'Internal command to provide shell completion suggestions';
+
+    private static $completionOutputs = [
+        'bash' => BashCompletionOutput::class,
+    ];
+    private $logger;
+
+    public function __construct(LoggerInterface $logger = null)
+    {
+        $this->logger = $logger;
+
+        parent::__construct();
+    }
+
+    protected function configure(): void
+    {
+        $this
+            ->setName(ltrim(self::$defaultName, '|'))
+            ->setDescription(self::$defaultDescription)
+            ->setHidden(true)
+            ->addOption('shell', 's', InputOption::VALUE_REQUIRED)
+            ->addOption('current', 'c', InputOption::VALUE_REQUIRED)
+            ->addOption('input', 'i', InputOption::VALUE_REQUIRED | InputOption::VALUE_IS_ARRAY)
+        ;
+    }
+
+    protected function execute(InputInterface $input, OutputInterface $output): int
+    {
+        try {
+            $shell = $input->getOption('shell');
+            if (!$shell) {
+                throw new \RuntimeException('The "--shell" option must be set.');
+            }
+
+            if (!$completionOutput = self::$completionOutputs[$shell] ?? false) {
+                throw new \RuntimeException(sprintf('Shell completion is not supported for your shell: "%s" (supported: "%s").', $shell, implode('", "', array_keys(self::$completionOutputs))));
+            }
+
+            $completionInput = $this->createCompletionInput($input);
+            $completion = new Completion();
+
+            $command = $this->findCommand($completionInput, $output);
+            if (null === $command) {
+                $this->getApplication()->complete($completionInput, $completion);
+            } elseif (
+                CompletionInput::TYPE_ARGUMENT_VALUE === $completionInput->getCompletionType()
+                && 'command' === $completionInput->getCompletionName()
+                && $command->getName() !== $completionInput->getCompletionValue()
+            ) {
+                $completion->suggestValue($command->getName());
+            } else {
+                $command->mergeApplicationDefinition();
+                $completionInput->bind($command->getDefinition());
+
+                $command->complete($completionInput, $completion);
+            }
+
+            $completionOutput = new $completionOutput();
+            $completionOutput->write($completion, $output);
+        } catch (\Throwable $e) {
+            if ($this->logger) {
```

Comment:
If your logger is set to STDOUT, you get the error as the completion here.

Reply:
Hmm, that's not great of course (I explicitly used this logger to avoid having the exception message as completion output).

What do you think about having a debug flag, like you proposed in https://github.com/symfony/symfony/pull/42251#issuecomment-893337446 and then always directly print the exception message to STDERR (at least, I expect this won't be picked up by compgen)?